File: packages/gbnf-experiment/data/20260909T012246Z_fbe56704/ported_implementation/gbnf/grammar_graph/get_parent_stack_id.py
from typing import TYPE_CHECKING, Callable, List, Union

from .colorize import Color

if TYPE_CHECKING:
    from .graph_pointer import GraphPointer


def get_parent_stack_id(
    pointer: "GraphPointer", col: Callable[[Union[str, int], Color], str]
) -> str:
    stack_ids: List[str] = []
    parent = pointer.parent
    while parent:
        meta = parent.node.meta
        stack_ids.append(f"{meta.stack_id},{meta.path_id},{meta.step_id}")
        parent = parent.parent
    return col("<-", Color.GRAY).join(col(id, Color.RED) for id in stack_ids)

File: proposed-projects/round-trip-experiment/reverse/20260909T153339Z_9de9fd66/20260910T090338Z_5eab4940/ported_implementation/src/grammar-graph/get-parent-stack-id.ts
import { Color, type Colorize } from './colorize.js';
import type { GraphPointer } from './graph-pointer.js';

export const getParentStackId = (pointer: GraphPointer, col: Colorize): string => {
  const stackIds: string[] = [];
  let parent = pointer.parent;
  while (parent) {
    const { stackId, pathId, stepId } = parent.node.meta;
    stackIds.push(`${stackId},${pathId},${stepId}`);
    parent = parent.parent;
  }
  return stackIds
    .map((stackId) => col(stackId, Color.RED))
    .join(col('<-', Color.GRAY));
};
